GoogleDataTransportCCTSupport/GDTCCTLibrary/GDTCCTPrioritizer.m

```diff
@@ -238,9 +257,127 @@ typedef NS_ENUM(NSInteger, GDTCCTQoSTier) {
+}
+
+- (void)encodeWithCoder:(NSCoder *)coder {
+  GDTCCTPrioritizer *sharedInstance = [GDTCCTPrioritizer sharedInstance];
+  NSMutableSet<GDTCOREvent *> *CCTEvents = sharedInstance->_CCTEvents;
+  if (CCTEvents) {
+    [coder encodeObject:CCTEvents forKey:GDTCCTUploaderCCTEventsKey];
+  }
+  NSMutableSet<GDTCOREvent *> *FLLEvents = sharedInstance->_FLLEvents;
+  if (FLLEvents) {
+    [coder encodeObject:FLLEvents forKey:GDTCCTUploaderFLLEventsKey];
+  }
+  NSMutableSet<GDTCOREvent *> *CSHEvents = sharedInstance->_CSHEvents;
+  if (CSHEvents) {
+    [coder encodeObject:CSHEvents forKey:GDTCCTUploaderCSHEventsKey];
+  }
+}
+
+#pragma mark - GDTCORLifecycleProtocol
+
+- (void)appWillForeground:(GDTCORApplication *)app {
+  if (@available(macOS 10.13, iOS 11.0, tvOS 11.0, *)) {
+    NSError *error;
+    NSData *data = [NSData dataWithContentsOfFile:ArchivePath()];
+    if (data) {
+      [NSKeyedUnarchiver unarchivedObjectOfClass:[GDTCCTPrioritizer class]
+                                        fromData:data
+                                           error:&error];
+    }
+  } else {
+#if !TARGET_OS_MACCATALYST && !TARGET_OS_WATCH
+    [NSKeyedUnarchiver unarchiveObjectWithFile:ArchivePath()];
+#endif
+  }
+}
+
+- (void)appWillBackground:(GDTCORApplication *)app {
+  dispatch_async(_queue, ^{
+    // Immediately request a background task to run until the end of the current queue of work, and
+    // cancel it once the work is done.
+    __block GDTCORBackgroundIdentifier bgID =
+        [app beginBackgroundTaskWithName:@"GDTStorage"
+                       expirationHandler:^{
+                         [app endBackgroundTask:bgID];
+                         bgID = GDTCORBackgroundIdentifierInvalid;
+                       }];
+
+    if (@available(macOS 10.13, iOS 11.0, tvOS 11.0, *)) {
+      NSError *error;
+      NSData *data = [NSKeyedArchiver archivedDataWithRootObject:self
+                                           requiringSecureCoding:YES
+                                                           error:&error];
+      [data writeToFile:ArchivePath() atomically:YES];
+    } else {
+#if !TARGET_OS_MACCATALYST && !TARGET_OS_WATCH
+      [NSKeyedArchiver archiveRootObject:self toFile:ArchivePath()];
+#endif
+    }
+
+    // End the background task if it's still valid.
+    [app endBackgroundTask:bgID];
+    bgID = GDTCORBackgroundIdentifierInvalid;
+  });
+}
+
+- (void)appWillTerminate:(GDTCORApplication *)application {
+  dispatch_sync(_queue, ^{
+    if (@available(macOS 10.13, iOS 11.0, tvOS 11.0, *)) {
+      NSError *error;
+      NSData *data = [NSKeyedArchiver archivedDataWithRootObject:self
+                                           requiringSecureCoding:YES
+                                                           error:&error];
+      [data writeToFile:ArchivePath() atomically:YES];
+    } else {
+#if !TARGET_OS_MACCATALYST && !TARGET_OS_WATCH
+      [NSKeyedArchiver archiveRootObject:self toFile:ArchivePath()];
```

Comment:
Secure coding is supported by previous OS versions. You may refer to [GoogleUtilities](https://github.com/firebase/firebase-ios-sdk/blob/dd70b46370dab27fb9e19853d586ee888c2c33b5/GoogleUtilities/Environment/GULSecureCoding.m#L30) as an example.

Reply:
I'm not sure I understand--this is still using secure coding, it's just avoiding calling a deprecated API on some platforms.

Reply:
Sorry, nm, mixed it with NSKeyedUnarchiver